staging/src/k8s.io/apiserver/pkg/admission/attributes.go

```diff
@@ -140,6 +145,73 @@ func (record *attributesRecord) AddAnnotation(key, value string) error {
 	return nil
 }
 
+func (record *attributesRecord) GetReinvocationContext() ReinvocationContext {
+	return record.reinvocationContext
+}
+
+func newReinvocationContext() *reinvocationContext {
+	return &reinvocationContext{reinvokeWebhooksIfNeeded: sets.NewString(), reinvokeWebhooks: sets.NewString()}
+}
+
+type reinvocationContext struct {
+	// isReinvoke is true when admission plugins are being reinvoked
+	isReinvoke bool
+	// lastOutput holds the result of the last webhook admission plugin call
+	lastOutput runtime.Object
+	// reinvokeWebhooksIfNeeded holds the set of webhooks to reinvoke if a later mutation occurs
+	reinvokeWebhooksIfNeeded sets.String
+	// reinvokeWebhooks holds the set of webhooks that should be reinvoked
+	reinvokeWebhooks sets.String
+	// reinvokeInTree indicates in-tree plugins should be reinvoked
+	reinvokeInTree bool
+}
+
+func (rc *reinvocationContext) IsReinvoke() bool {
+	return rc.isReinvoke
+}
+
+func (rc *reinvocationContext) SetIsReinvoke() {
+	rc.isReinvoke = true
+}
+
+func (rc *reinvocationContext) AddReinvokeIfNeeded(webhook string) {
+	rc.reinvokeWebhooksIfNeeded.Insert(webhook)
+}
+
+func (rc *reinvocationContext) ShouldReinvoke() bool {
+	return rc.reinvokeInTree || len(rc.reinvokeWebhooks) > 0
+}
+
+func (rc *reinvocationContext) CheckIfLastOutputChanged(object runtime.Object) {
+	if rc.isReinvoke && !reflect.DeepEqual(rc.lastOutput, object) {
+		// If the object has changed, we know the in-tree plugin reinvocations have mutated the object,
+		// and we need to reinvoke all eligible webhooks.
+		rc.prepareToReinvokeAllIfNeededWebhooks()
+	}
+}
+
+func (rc *reinvocationContext) SetLastOutput(object runtime.Object) {
+	rc.lastOutput = object.DeepCopyObject()
+}
+
+func (rc *reinvocationContext) ShouldInvokeWebhook(webhook string) bool {
+	return !rc.isReinvoke || rc.reinvokeWebhooks.Has(webhook)
+}
+
+func (rc *reinvocationContext) RecordWebhookMutation(webhook string) {
```

Comment:
remove the unused input

I'd also suggest rename the function to something like "RequireReinvokingPreviousPlugins" and document that "the function is called when a mutating webhook results in a mutation, we require all previous applicable plugins in chain (built-in plugins or webhooks) to be re-invoked"

Reply:
Yes, this could be much clearer. I've done a pass over the interface to incorporate this and generally clarify the purpose of the various functions.